yellow-paper/docs/addresses-and-keys/addresses-and-keys.md

```diff
@@ -0,0 +1,903 @@
+
+Requirements:
+- A tx authentication secret key SHOULD NOT enter Aztec software, and SHOULD NOT enter a circuit.
+  - Reason: this is just best practice.
+
+
+### Master & Siloed Keys
+
+Requirements:
+- All keys must be re-derivable from a single `seed` secret.
+- Users must have the option of keeping this `seed` offline, e.g. in a hardware wallet, or on a piece of paper.
+- All master keys (for a particular user) must be linkable to a single "user identifier" for that user.
+  - Notice: we don't prescribe whether this "user identifier" must be a public key, or a field (i.e. an address). The below protocol suggestion has ended up with an address, but that's not to say it's required.
+- For each contract, a siloed set of all keys MUST be derivable.
+  - Reason: secret keys must be siloed, so that a malicious app circuit cannot access and emit (as an unencrypted event or as args to a public function) a user's master secret keys or the secret keys of other apps.
+- Master _secret_ keys must not be passed into an app circuit.
+  - Reason: a malicious app could broadcast these secret keys to the world.
+- Siloed secret keys _of other apps_ must not be passed into an app circuit.
+  - Reason: a malicious app could broadcast these secret keys to the world.
+- The PXE must prevent an app from accessing master secret keys.
+- The PXE must prevent an app from accessing siloed secret keys that belong to another contract address.
+  - Note: To achieve this, the PXE simulator will need to check whether the bytecode being executed (that is requesting secret keys) actually exists at the contract address.
+- There must be one and only one way to derive all (current*) master keys, and all siloed keys, for a particular "user identifier".
+  - For example, a user should not be able to derive multiple different outgoing viewing keys for a single incoming viewing key (note: this was a 'bug' that was fixed between ZCash Sapling and Orchard).
+  - \*"current", alludes to the possibility that the protocol might wish to support rotating of keys, but only if one and only one set of keys is derivable as "current".
+- All app-siloed keys can all be deterministically linked back to the user's address, without leaking important secrets to the app.
+
+### Account Contract Bytecode
+
+Requirements:
+- The (current) bytecode of an account contract must be discoverable, based on the "user identifier" (probably an address).
+  - Reason: Although in 'private land', the king of the hill griefing problem prevents private account contract acir from being callable, there might still be uses where a _public_ function of an account contract might wish to be called by someone.
+- The (current) bytecode of an account contract must therefore be 'tied to' to the "user identifier".
+- The constructor arguments of an account contract must be discoverable.
```

Comment:
I think the constructor arguments of every contract that others might interact with must be discoverable, so that people can validate the initial state variables of the contract.
Maybe "discoverable" is the wrong term. Maybe "The constructor arguments that were used to deploy an account contract must be verifiable."

Reply:
But doesn't the private kernel proof for the execution of the constructor verify this? It's proving that some private constructor args were provided, such that the resulting state was X.

Reply:
I think a user might need to verify some initial state for themselves.
E.g. a token contract, where someone claims "I set the initial supply to 1m tokens", when in fact they set it to be 1bn tokens. (I suppose it's a silly example, because the user could query a 'get_token_supply' function, but I still think it's important.
E.g. an account contract, where a merkle tree whose leaves are multisig public keys is initialised. Then it's important for all multisig participants to be able to verify which public keys were passed into the constructor, to build the tree (to ensure no cheeky extra keys were included, for example). 

I think with this requirement, I was trying to reverse-justify why we embed the constructor arguments into the address.